KAFKA-8602: Fix bug in stand-by task creation

cadonna opened the pull request:
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

cadonna committed:
KAFKA-8602: Fix bug in stand-by task creation

cadonna commented:
Call for review: @mjsax @bbejeck @vvcephei @abbccdda @guozhangwang @ableegoldman 
The fix needs to be cherry-picked back to 1.0.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.NUM_STANDBY_REPLICAS_CONFIG, 1);
+        streamsConfiguration.put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 1);
```
Nit: no need to overwrite; `1` is the default anyway

cadonna replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.NUM_STANDBY_REPLICAS_CONFIG, 1);
+        streamsConfiguration.put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 1);
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
```
as above

cadonna replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    @Test
+    public void shouldNotCreateAnyStandByTasksForStateStoreWithLoggingDisabled() throws InterruptedException {
```
nit: simplify `InterruptedException` -> `Exception`

cadonna replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+
+                @Override
+                public KeyValue<Integer, Integer> transform(final Integer key, final Integer value) {
+                    state.putIfAbsent(key, value);
```
We we actually need to use the store? Seems we can remove this code?

cadonna replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+    @BeforeClass
+    public static void createTopics() throws InterruptedException {
+        CLUSTER.createTopic(INPUT_TOPIC, 2, 1);
+        CLUSTER.createTopic(OUTPUT_TOPIC, 2, 1);
```
Why do we need an output topic? We never use it to consume a result?

cadonna replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+                public KeyValue<Integer, Integer> transform(final Integer key, final Integer value) {
+                    state.putIfAbsent(key, value);
+                    final KeyValue<Integer, Integer> result = new KeyValue<>(key, value);
+                    return result;
```
Can just return `null` ?

cadonna replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+        final KafkaStreams client1 = new KafkaStreams(topology, streamsConfiguration());
+        final KafkaStreams client2 = new KafkaStreams(topology, streamsConfiguration());
+
+        final boolean[] client1IsOk = {false}; // has to be a final array, otherwise flag cannot be modified in lambda
```
nit: remove comment (we would make it final in any case anyway).

I am wondering if it should be `volatile` thought, as the state listener callback is used by a different thread?

cadonna replied:
Here the main point of the comment is that we need to use an array of booleans instead of a plain boolean variable, because otherwise I cannot declare it final and modify it within the lambda.

Since I used a volatile variable now (good point, btw), the array is not needed anymore. Hence, I removed the comment.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+                client1IsOk[0] = true;
+            }
+        });
+        final boolean[] client2IsOk = {false}; // has to be a final array, otherwise flag cannot be modified in lambda
```
as above.

cadonna replied:
Ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/StandByTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,139 @@
+        TestUtils.waitForCondition(
+            () -> client1IsOk[0] && client2IsOk[0],
+            30 * 1000,
+            "At least one client is not in state RUNNING or has a stand-by task");
```
Should we split both conditions? Maybe only wait for state `RUNNING` and check `localThreadsMetadata` after the wait condition?

cadonna replied:
I tried it and the test became flaky. The reason is that without fix the client is in `RUNNING` when the `IllegalStateException` is thrown, then it changes to `ERROR`, `PENDING_SHUTDOWN`, and finally `NOT_RUNNING`. It could happen that the wait condition gets satisfied when the client is in RUNNING. Then when the tests verifies `localThreadMetadata` two scenario can occur: 

1) client is still in RUNNING: `localThreadMetadata` contains a stand-by task -> assertion not satisfied 
2) client is not in RUNNING: exception is thrown because `localThreadMetadata` can only be accessed when the client is in `RUNNING`.

Both scenarios would let the test fail which would be OK, but not really clean. However, the test was sometimes green, which I currently do not understand. My guess would be race condition.

My approach checks 'localThreadMetadata' when the state changes to `RUNNING`, which should be safe in the error case. In the good case the test could run into the timeout if start-up is slow, though.

mjsax reviewed (commented):
Overall LGTM. Couple of minor comments.

cadonna committed:
Incorporate comments

cadonna committed:
Merge remote-tracking branch 'upstream/trunk' into AK8602-bug_standby_task_creation

cadonna commented:
Retest this, please

cadonna committed:
Merge remote-tracking branch 'upstream/trunk' into AK8602-bug_standby_task_creation

cadonna committed:
Verify all stream threads of a client

- Verifies that all stream threads of a client do not have assigned
  stand-by tasks. Before only the first stream thread returned by the iterator
  was verified. Now the test is safe to be run with multiple stream threads
  per client.
- Renamed test class to conform with format of `StandbyTask` class

cadonna committed:
Fix checkstyle errors

cadonna committed:
Merge remote-tracking branch 'upstream/trunk' into AK8602-bug_standby_task_creation

cadonna commented:
failures unrelated

Retest this, please

cadonna commented:
Retest this, please

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,134 @@
+        client1.setStateListener((newState, oldState) -> {
+            if (newState == State.RUNNING &&
+                client1.localThreadsMetadata().stream().allMatch(thread -> thread.standbyTasks().isEmpty())) {
+
```
nit: extra line

cadonna replied:
I left this extra line on purpose so that it is immediately clear that `client1IsOk = true;` is not part of the `if` condition. Since indentation of line 113 and 115 is both four characters, both lines would appear at first sight as if they belonged to same code block even though they do not.

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskCreationIntegrationTest.java:
```diff
@@ -0,0 +1,134 @@
+        TestUtils.waitForCondition(
+            () -> client1IsOk && client2IsOk,
+            30 * 1000,
+            "At least one client did not reach state RUNNING without any stand-by tasks");
```
We could actually show the client state by:
`"Some clients didn't reach state RUNNING without any stand-by tasks. Eventual status: [Client1: {}, Client2: {}]", client1IsOk, client2IsOk`

cadonna replied:
Ack

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1057,6 +1063,69 @@ public void shouldUpdateStandbyTask() throws Exception {
 
+    @Test
+    public void shouldCreateStandbyTask() {
+        final MockProcessor mockProcessor = new MockProcessor();
```
We could reuse initialization code for L1068 - L1070

cadonna replied:
Ack

abbccdda commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1057,6 +1063,69 @@ public void shouldUpdateStandbyTask() throws Exception {
+        internalTopologyBuilder.addStateStore(new MockKeyValueStoreBuilder("myStore", true), "processor1");
+        final StreamThread.StandbyTaskCreator standbyTaskCreator = createStandbyTaskCreator(internalTopologyBuilder);
+
+        final StandbyTask standbyTask = standbyTaskCreator.createTask(
```
The standby task could also be reused IIUC

cadonna replied:
Ack

cadonna committed:
Merge remote-tracking branch 'upstream/trunk' into AK8602-bug_standby_task_creation

cadonna committed:
Incorporate review comments

cadonna commented:
Retest this, please

cadonna commented:
Retest this, please

cadonna commented:
failures unrelated

Retest this, please

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -495,7 +495,7 @@ StandbyTask createTask(final Consumer<byte[], byte[]> consumer,
             final ProcessorTopology topology = builder.build(taskId.topicGroupId);
 
-            if (!topology.stateStores().isEmpty()) {
+            if (!topology.stateStores().isEmpty() && !topology.storeToChangelogTopic().isEmpty()) {
```
Does this still work for optimized source tables, which read from the input topic instead?

mjsax replied:
Good point. Thinking about this, StandBys for source-KTables might have been broker for a long time already... (maybe since 0.10.0.0???)

Maybe @cadonna can verify? If that is the case, we should split out a separate ticket and PR to fix StandBys for source-KTables independently.

cadonna replied:
Yeah, good point! Added an integration test to verify materialized and optimized source tables.

mjsax replied:
Cannot follow. You test seem to use the PAPI, and the PAPI does not provide the KTable optimization. You would need to use `StreamBuilder#table()` to test the changelog optimization.

cadonna replied:
See `StandbyTaskCreationIntegrationTest` line 128.

vvcephei reviewed (commented):
Hey @cadonna , thanks for the fix! Just one quick question...

cadonna committed:
Merge remote-tracking branch 'upstream/trunk' into AK8602-bug_standby_task_creation

cadonna committed:
Merge remote-tracking branch 'upstream/trunk' into AK8602-bug_standby_task_creation

cadonna committed:
Add integration test for materialized and optimized source table

cadonna commented:
Retest this, please

cadonna commented:
Failures unrelated
Retest this, please

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskCreationIntegrationTest.java:
```diff
@@ -105,32 +106,81 @@ public void close() {}
+        final Properties streamsConfiguration1 = streamsConfiguration();
+        streamsConfiguration1.put(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE);
+        final Properties streamsConfiguration2 = streamsConfiguration();
+        streamsConfiguration2.put(StreamsConfig.TOPOLOGY_OPTIMIZATION, StreamsConfig.OPTIMIZE);
```
Probably not that important, but are these two Properties objects exactly the same?

cadonna replied:
No, they are not because of `streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory(applicationId).getPath());`. Each instance of the `Properties` has a different state directory, which is good because the test does not work otherwise. I tried it out.

vvcephei reviewed (approved):
Thanks, @cadonna !

bbejeck reviewed (approved):
Thanks for the fix @cadonna LGTM

bbejeck commented:
retest this please

bbejeck commented:
Retest this, please

cadonna commented:
Failure unrelated 
Retest this, please

cadonna commented:
Retest this, please

cadonna commented:
Retest this, please

cadonna commented:
Retest this, please

cadonna commented:
Reopen to trigger build

cadonna commented:
Retest this, please

bbejeck commented:
retest this please

bbejeck commented:
3 green builds, failing 4th build is from disabling the previous GitHub PR builder in favor of a new one.  There were some errors with the new PR builder, so the previous GitHub PR builder has been re-enabled, so merging this PR.

bbejeck commented:
Merged #7008 into trunk

guozhangwang commented:
retest this please

bbejeck commented:
cherry-picked to 2.3 via #7092 
and 7092 was cherry-picked to 2.2 and 2.1
